packages/statemanager/src/ethersStateManager.ts

```diff
@@ -0,0 +1,253 @@
+
+    const storageTrie = await this._lookupStorageTrie(address)
+    const foundValue = await storageTrie.get(key) //@ts-ignore
+
+    return Buffer.from(RLP.decode(Uint8Array.from(foundValue ?? [])) as Uint8Array)
+  }
+
+  async getContractStorageFromProvider(address: Address, key: Buffer): Promise<void> {
+    const accountData = await this.provider.send('eth_getProof', [
+      address.toString(),
+      [bufferToHex(key)],
+      this.blockTag,
+    ])
+    const rawData = accountData.accountProof
+    await this.trie.fromProof(rawData.map((e: string) => hexToBytes(e)))
+    // Only requesting a single slot at a time so the proof will always be the first item in the array
+    const storageData = accountData.storageProof[0]
+
+    await this.trie.fromProof(storageData.proof.map((e: string) => hexToBytes(e)))
+  }
+  async putContractStorage(address: Address, key: Buffer, value: Buffer): Promise<void> {
+    const storageTrie = await this._lookupStorageTrie(address)
+    await storageTrie.put(key, arrToBufArr(RLP.encode(value)))
+    const contract = await this.getAccount(address)
+    contract.storageRoot = storageTrie.root()
+    await this.putAccount(address, contract)
+  }
+
+  copy(): EthersStateManager {
+    return new EthersStateManager({ provider: this.provider })
+  }
+
+  dumpStorage(_address: Address): Promise<StorageDump> {
+    throw new Error('Method not implemented.')
+  }
+
+  async accountExists(address: Address): Promise<boolean> {
+    log(`Verify if ${address.toString()} exists`)
+    const account = this._cache.get(address)
+    if (account.isEmpty()) {
+      // Get latest block (or block specified in `this.blockTag`)
+      const block = await this.provider.send('eth_getBlockByNumber', [this.blockTag, false])
+      // Get merkle proof for `address` from provider
+      const proof = await this.provider.send('eth_getProof', [address.toString(), [], block.number])
+      const proofBuf = proof.accountProof.map((proofNode: string) => toBuffer(proofNode))
+      const trie = new Trie({ useKeyHashing: true })
+      const verified = await trie.verifyProof(toBuffer(block.stateRoot), address.buf, proofBuf)
+      // if not verified (i.e. verifyProof returns null), account does not exist
+      if (verified === null) return false
+    }
+    return true
+  }
+
+  async getAccount(address: Address): Promise<Account> {
+    const account = await this._cache.getOrLoad(address)
+    return account
+  }
+  async getAccountFromProvider(address: Address): Promise<Account> {
+    log(`Retrieving account data for ${address.toString()} from provider`)
+
+    const accountData = await this.provider.send('eth_getProof', [
+      address.toString(),
+      [],
+      this.blockTag,
+    ])
+    const rawData = accountData.accountProof
+
+    await this.trie.fromProof(rawData.map((e: string) => hexToBytes(e)))
+
+    const account = Account.fromAccountData({
```

Comment:
Does this work if we have 2 txs from the same account in the block? Does it save the account in cache, or does it get this from the external provider again and thus overrides the earlier operation?

Reply:
It should never be called twice for the same account.  `getAccountFromProvider` is called within `getAccount` which checks the cache first and only calls out to the provider if the account isn't found.

Reply:
i think two txs refering the same account should work, because first time it fetches and writes to cache, and if there are modifications, that are written to the cache, so for the next tx, it will correctly get the modified account